kube/src/config/auth.rs

```diff
@@ -0,0 +1,392 @@
+    pub(crate) async fn from_auth_info(auth_info: &AuthInfo) -> Result<Self> {
+        if let Some(provider) = &auth_info.auth_provider {
+            match token_from_provider(provider).await? {
+                (token, Some(expiry)) => {
+                    let mut info = auth_info.clone();
+                    let mut provider = provider.clone();
+                    provider.config.insert("access-token".into(), token.clone());
+                    provider.config.insert("expiry".into(), expiry.to_rfc3339());
+                    info.auth_provider = Some(provider);
+                    return Ok(Self::RefreshableToken(Arc::new(Mutex::new((
+                        format!("Bearer {}", token),
+                        expiry,
+                        info,
+                    )))));
+                }
+
+                (token, None) => {
+                    return Ok(Self::Token(format!("Bearer {}", token)));
+                }
+            }
+        }
+
+        let (raw_token, expiration) = match &auth_info.token {
+            Some(token) => (Some(token.clone()), None),
+            None => {
+                if let Some(exec) = &auth_info.exec {
+                    let creds = auth_exec(exec)?;
+                    let status = creds.status.ok_or(ConfigError::ExecPluginFailed)?;
+                    let expiration = match status.expiration_timestamp {
+                        Some(ts) => Some(
+                            ts.parse::<DateTime<Utc>>()
+                                .map_err(ConfigError::MalformedTokenExpirationDate)?,
+                        ),
+                        None => None,
+                    };
+                    (status.token, expiration)
+                } else {
+                    (None, None)
+                }
+            }
+        };
+
+        match (
+            utils::data_or_file(&raw_token, &auth_info.token_file),
+            (&auth_info.username, &auth_info.password),
+            expiration,
+        ) {
+            (Ok(token), _, None) => Ok(Authentication::Token(format!("Bearer {}", token))),
+            (Ok(token), _, Some(expire)) => Ok(Authentication::RefreshableToken(Arc::new(Mutex::new((
+                format!("Bearer {}", token),
+                expire,
+                auth_info.clone(),
+            ))))),
+            (_, (Some(u), Some(p)), _) => {
+                let encoded = base64::encode(&format!("{}:{}", u, p));
+                Ok(Authentication::Basic(format!("Basic {}", encoded)))
+            }
+            _ => Ok(Authentication::None),
+        }
+    }
+}
+
+async fn token_from_provider(provider: &AuthProviderConfig) -> Result<(String, Option<DateTime<Utc>>)> {
+    if let Some(id_token) = provider.config.get("id-token") {
+        return Ok((id_token.clone(), None));
+    }
+
+    if let Some(access_token) = provider.config.get("access-token") {
+        if let Some(expiry) = provider.config.get("expiry") {
```

Comment:
I guess `expiry` is GCP only? https://github.com/clux/kube-rs/issues/238 was most likely caused by

https://github.com/clux/kube-rs/blob/d450509554ef3f601a0e2ba90524b9fae43c1a7b/kube/src/config/file_config.rs#L173

`token_from_provider` needs to check `provider.name` first then.

Reply:
From a quick skim of the [auth providers](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/client-go/plugin/pkg/client/auth), I can only see the `expiry` field in [gcp](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/plugin/pkg/client/auth/gcp/gcp.go), not [azure](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/plugin/pkg/client/auth/azure/azure.go) at least.

Reply:
yeah, looks like you're right, need to be more defensive there for other auth providers. the people who provided the oauth modules were pretty much all using gcp.